packages/fuels-core/src/lib.rs

```diff
@@ -273,6 +283,60 @@ impl Tokenizable for u64 {
     }
 }
 
+// Here we implement `Tokenizable` for a given tuple of a given length.
+// This is done this way because we can't use `impl<T> Tokenizable for (T,)`.
+// So we implement `Tokenizable` for each tuple length, covering
+// a reasonable range of tuple lengths.
+macro_rules! impl_tuples {
+    ($num: expr, $( $ty: ident : $no: tt, )+) => {
+        impl<$($ty, )+> Tokenizable for ($($ty,)+) where
+            $(
+                $ty: Tokenizable,
+            )+
+        {
+            fn from_token(token: Token) -> Result<Self, InvalidOutputType> {
+                match token {
+                    Token::Tuple(mut tokens) => {
+                        let mut it = tokens.drain(..);
+                        Ok(($(
+                          $ty::from_token(it.next().expect("All elements are in vector."))?,
+                        )+))
+                    },
+                    other => Err(InvalidOutputType(format!(
+                        "Expected `Tuple`, got {:?}",
+                        other,
+                    ))),
+                }
+            }
+
+            fn into_token(self) -> Token {
+                Token::Tuple(vec![
+                    $( self.$no.into_token(), )+
+                ])
+            }
+        }
+    }
+}
```

Comment:
I think for clarity, we should maybe include an example of the macro's expansion as doc somewhere (eg for tuples of len 2), because reading this and checking its validity is pretty hard imo.

Reply:
Here's the generated impl for a tuple of len 2:

```Rust
impl<A,B, >Tokenizable for(A,B,)where A:Tokenizable,B:Tokenizable,{
  fn from_token(token:Token) -> Result<Self,InvalidOutputType>{
    match token {
      Token::Tuple(mut tokens) => {
        let mut it = tokens.drain(..);
        Ok((A::from_token(it.next().expect("All elements are in vector."))?,B::from_token(it.next().expect("All elements are in vector."))?,))
      },
      other => Err(InvalidOutputType({
        let res = $crate::fmt::format(unsafe {
          std::fmt::Arguments::new_v1(&[], &[std::fmt::ArgumentV1::new(&(other),std::fmt::Display::fmt),])
        });
        res
      })),
    
      }
  }
  fn into_token(self) -> Token {
    Token::Tuple((<[_]>::into_vec(box[(self.0.into_token()),(self.1.into_token())])))
  }
}
```

If you tell me that this is a lot different/more readable from/than the macro definition:

```Rust
macro_rules! impl_tuples {
    ($num: expr, $( $ty: ident : $no: tt, )+) => {
        impl<$($ty, )+> Tokenizable for ($($ty,)+) where
            $(
                $ty: Tokenizable,
            )+
        {
            fn from_token(token: Token) -> Result<Self, InvalidOutputType> {
                match token {
                    Token::Tuple(mut tokens) => {
                        let mut it = tokens.drain(..);
                        Ok(($(
                          $ty::from_token(it.next().expect("All elements are in vector."))?,
                        )+))
                    },
                    other => Err(InvalidOutputType(format!(
                        "Expected `Tuple`, got {:?}",
                        other,
                    ))),
                }
            }

            fn into_token(self) -> Token {
                Token::Tuple(vec![
                    $( self.$no.into_token(), )+
                ])
            }
        }
    }
}
```

Or the `impl Tokenizable for Vec<T>` right above https://github.com/FuelLabs/fuels-rs/blob/fe6727e74d52cf7fd2118a23642a28638133eacd/packages/fuels-core/src/lib.rs#L211-L232 

Which is very similar to the generated one for tuples...

Then yes, I'll 20 extra lines of documentation code to clarify it 😛.

Reply:
Fair enough, it's not more readable! :stuck_out_tongue_closed_eyes: